controller/scheduler-v2/scheduler.go

```diff
@@ -145,12 +154,69 @@ func (s *Scheduler) Sync() (err error) {
 					continue
 				}
 				log.Info("adding job")
-				s.AddJob(NewJob(f, jobType, h.ID(), job.ID), appName, utils.JobMetaFromMetadata(job.Metadata))
+				j := NewJob(f, jobType, h.ID(), job.ID)
+				s.AddJob(j, appName, utils.JobMetaFromMetadata(job.Metadata))
+				fc.AddJob(j)
 			}
 		}
 	}
-	// TODO rectify formations
-	return
+	for jobID := range unSyncedJobs {
+		delete(s.jobs, jobID)
+	}
+	return s.rectifyFormations(fc)
+}
+
+func (s *Scheduler) SyncFormations() (err error) {
+	log := s.log.New("fn", "SyncFormations")
+
+	log.Info("getting formations")
+
+	if len(s.formations) == 0 {
+		log.Info("getting apps")
+		apps, err := s.AppList()
+		if err != nil {
+			return err
+		}
+		for _, app := range apps {
+			fs, err := s.FormationList(app.ID)
+			if err != nil {
+				return err
+			}
+			for _, f := range fs {
+				form, err := s.getFormation(app.ID, app.Name, f.ReleaseID)
+				if err != nil {
+					return err
+				}
+				s.formations[form.key()] = form
+			}
+		}
+	}
+	return err
+}
+
+func (s *Scheduler) rectifyFormations(fc formationCounts) (err error) {
+	for fKey, formation := range s.formations {
+		for typ, count := range formation.Processes {
+			diff := count - fc[fKey][typ]
+
+			if diff > 0 {
+				for i := 0; i < diff; i++ {
+					s.jobRequests <- NewJobRequest(formation, JobRequestTypeUp, typ, "", "")
+				}
+			} else if diff < 0 {
+				for i := 0; i < -diff; i++ {
+					s.jobRequests <- NewJobRequest(formation, JobRequestTypeDown, typ, "", "")
+				}
+			}
+			delete(fc[fKey], typ)
+		}
+		if len(fc[fKey]) == 0 {
+			delete(fc, fKey)
+		} else {
+			err = fmt.Errorf("Unknown job type in formation %v", fc[fKey])
+		}
+	}
+	return err
```

Comment:
The rectifying of formations after `SyncFormations` needs to update the in-memory formations so that we have the latest expected process counts in memory, so this should probably just trigger a `FormationChange` for each formation retrieved in the sync.